File: app/src/main/java/com/haoyu/app/fragment/PageAllQuestionFragment.java
package com.haoyu.app.fragment;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.view.View;
import android.widget.TextView;

import com.haoyu.app.activity.AppQuestionDetailActivity;
import com.haoyu.app.activity.AppQuestionEditActivity;
import com.haoyu.app.adapter.PageQuestionAdapter;
import com.haoyu.app.base.BaseFragment;
import com.haoyu.app.base.BaseResponseResult;
import com.haoyu.app.basehelper.BaseRecyclerAdapter;
import com.haoyu.app.entity.FAQsEntity;
import com.haoyu.app.entity.FAQsListResult;
import com.haoyu.app.entity.FollowMobileEntity;
import com.haoyu.app.entity.FollowMobileResult;
import com.haoyu.app.entity.Paginator;
import com.haoyu.app.lego.student.R;
import com.haoyu.app.rxBus.MessageEvent;
import com.haoyu.app.rxBus.RxBus;
import com.haoyu.app.utils.Action;
import com.haoyu.app.utils.Constants;
import com.haoyu.app.utils.OkHttpClientManager;
import com.haoyu.app.view.LoadFailView;
import com.haoyu.app.view.LoadingView;
import com.haoyu.app.xrecyclerview.XRecyclerView;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import butterknife.BindView;
import okhttp3.Request;

/**
 * 创建日期：2017/8/16 on 9:59
 * 描述:
 * 作者:马飞奔 Administrator
 */
public class PageAllQuestionFragment extends BaseFragment implements XRecyclerView.LoadingListener {
    @BindView(R.id.loadingView)
    LoadingView loadingView;
    @BindView(R.id.loadFailView)
    LoadFailView loadFailView;
    @BindView(R.id.xRecyclerView)
    XRecyclerView xRecyclerView;
    @BindView(R.id.emptyView)
    TextView emptyView;
    private List<FAQsEntity> mDatas = new ArrayList<>();
    private PageQuestionAdapter adapter;
    private String type, relationId, relationType;
    private int page = 1;
    private boolean isRefresh, isLoadMore;
    private OnResponseListener onResponseListener;
    private int questionIndex = -1;
    private final int ANSWER_CODE = 10;

    public void setOnResponseListener(OnResponseListener onResponseListener) {
        this.onResponseListener = onResponseListener;
    }

    @Override
    public int createView() {
        return R.layout.fragment_page_question_child;
    }

    @Override
    public void initView(View view) {
        Bundle bundle = getArguments();
        if (bundle != null) {
            type = bundle.getString("type");
            relationId = bundle.getString("relationId");
            relationType = bundle.getString("relationType");
        }
        LinearLayoutManager layoutManager = new LinearLayoutManager(context);
        layoutManager.setOrientation(LinearLayoutManager.VERTICAL);
        xRecyclerView.setLayoutManager(layoutManager);
        if (type != null && type.equals("course"))
            adapter = new PageQuestionAdapter(context, mDatas, 1);
        else
            adapter = new PageQuestionAdapter(context, mDatas, 2);
        xRecyclerView.setAdapter(adapter);
        xRecyclerView.setLoadingListener(this);
        emptyView.setText(getResources().getString(R.string.empty_questions));
    }

    @Override
    public void initData() {
        String url = Constants.OUTRT_NET + "/m/faq_question" + "?relation.id=" + relationId + "&relation.type=" + relationType + "&page=" + page + "&orders=CREATE_TIME.DESC";
        addSubscription(OkHttpClientManager.getAsyn(context, url, new OkHttpClientManager.ResultCallback<FAQsListResult>() {
            @Override
            public void onBefore(Request request) {
                if (isRefresh || isLoadMore) {
                    loadingView.setVisibility(View.GONE);
                } else {
                    loadingView.setVisibility(View.VISIBLE);
                }
            }

            @Override
            public void onError(Request request, Exception e) {
                loadingView.setVisibility(View.GONE);
                if (isRefresh) {
                    xRecyclerView.refreshComplete(false);
                } else if (isLoadMore) {
                    page -= 1;
                    xRecyclerView.loadMoreComplete(false);
                } else {
                    xRecyclerView.setVisibility(View.GONE);
                    loadFailView.setVisibility(View.VISIBLE);
                }
            }

            @Override
            public void onResponse(FAQsListResult response) {
                loadingView.setVisibility(View.GONE);
                if (response != null && response.getResponseData() != null && response.getResponseData().getQuestions() != null && response.getResponseData().getQuestions().size() > 0) {
                    updateUI(response.getResponseData().getQuestions(), response.getResponseData().getPaginator());
                } else {
                    if (isRefresh) {
                        xRecyclerView.refreshComplete(true);
                    } else if (isLoadMore) {
                        xRecyclerView.loadMoreComplete(true);
                    } else {
                        xRecyclerView.setVisibility(View.GONE);
                        emptyView.setVisibility(View.VISIBLE);
                    }
                    xRecyclerView.setLoadingMoreEnabled(false);
                }
            }
        }));
    }

    private void updateUI(List<FAQsEntity> list, Paginator paginator) {
        if (xRecyclerView.getVisibility() != View.VISIBLE)
            xRecyclerView.setVisibility(View.VISIBLE);
        if (isRefresh) {
            mDatas.clear();
            xRecyclerView.refreshComplete(true);
        } else if (isLoadMore) {
            xRecyclerView.loadMoreComplete(true);
        }
        mDatas.addAll(list);
        adapter.notifyDataSetChanged();
        if (paginator != null && paginator.getHasNextPage()) {
            xRecyclerView.setLoadingMoreEnabled(true);
        } else {
            xRecyclerView.setLoadingMoreEnabled(false);
        }
        if (onResponseListener != null) {
            if (paginator != null)
                onResponseListener.getTotalCount(paginator.getTotalCount());
            else
                onResponseListener.getTotalCount(mDatas.size());
        }
    }

    @Override
    public void setListener() {
        adapter.setCollectCallBack(new PageQuestionAdapter.CollectCallBack() {
            @Override
            public void collect(int position, FAQsEntity entity) {
                collection(position);
            }

            @Override
            public void cancelCollect(int position, FAQsEntity entity) {
                cancelCollection(position);
            }
        });
        adapter.setAnswerCallBack(new PageQuestionAdapter.AnswerCallBack() {
            @Override
            public void answer(int position, FAQsEntity entity) {
                questionIndex = position;
                Intent intent = new Intent();
                intent.setClass(context, AppQuestionEditActivity.class);
                intent.putExtra("isAnswer", true);
                intent.putExtra("questionId", entity.getId());
                startActivityForResult(intent, ANSWER_CODE);
            }
        });
        adapter.setOnItemClickListener(new BaseRecyclerAdapter.OnItemClickListener() {
            @Override
            public void onItemClick(BaseRecyclerAdapter adapter, BaseRecyclerAdapter.RecyclerHolder holder, View view, int position) {
                if (position - 1 >= 0 && position - 1 < mDatas.size()) {
                    questionIndex = position - 1;
                    Intent intent = new Intent();
                    intent.setClass(context, AppQuestionDetailActivity.class);
                    intent.putExtra("type", type);
                    intent.putExtra("entity", mDatas.get(questionIndex));
                    startActivity(intent);
                }
            }
        });
    }

    /**
     * 取消收藏
     *
     * @param position
     */
    private void cancelCollection(final int position) {
        FollowMobileEntity follow = mDatas.get(position).getFollow();
        if (follow != null) {
            String url = Constants.OUTRT_NET + "/m/follow/" + follow.getId();
            Map<String, String> map = new HashMap<>();
            map.put("_method", "delete");
            addSubscription(OkHttpClientManager.postAsyn(context, url, new OkHttpClientManager.ResultCallback<BaseResponseResult>() {
                @Override
                public void onError(Request request, Exception exception) {
                    onNetWorkError();
                }

                @Override
                public void onResponse(BaseResponseResult response) {
                    if (response != null && response.getResponseCode() != null && response.getResponseCode().equals("00")) {
                        mDatas.get(position).setFollow(null);
                        adapter.notifyDataSetChanged();
                        MessageEvent event = new MessageEvent();
                        event.action = Action.COLLECTION;
                        event.obj = mDatas.get(position);
                        RxBus.getDefault().post(event);
                    }
                }
            }, map));
        }
    }

    /**
     * 创建收藏
     *
     * @param position
     */
    private void collection(final int position) {
        String url = Constants.OUTRT_NET + "/m/follow";
        Map<String, String> map = new HashMap<>();
        map.put("followEntity.id", mDatas.get(position).getId());
        map.put("followEntity.type", "course_study_question");
        OkHttpClientManager.postAsyn(context, url, new OkHttpClientManager.ResultCallback<FollowMobileResult>() {
            @Override
            public void onError(Request request, Exception exception) {
                onNetWorkError();
            }

            @Override
            public void onResponse(FollowMobileResult response) {
                if (response != null && response.getResponseCode() != null && response.getResponseCode().equals("00")) {
                    FollowMobileEntity entity = response.getResponseData();
                    mDatas.get(position).setFollow(entity);
                    adapter.notifyDataSetChanged();
                    MessageEvent event = new MessageEvent();
                    event.action = Action.COLLECTION;
                    event.obj = mDatas.get(position);
                    RxBus.getDefault().post(event);
                } else {
                    if (response.getResponseMsg() != null) {
                        toast(response.getResponseMsg());
                    }
                }
            }
        }, map);
    }

    @Override
    public void onRefresh() {
        isRefresh = true;
        isLoadMore = false;
        page = 1;
        initData();
    }

    @Override
    public void onLoadMore() {
        isRefresh = false;
        isLoadMore = true;
        page += 1;
        initData();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, requestCode, data);
        if (requestCode == ANSWER_CODE && resultCode == Activity.RESULT_OK && data != null) {
            int faqAnswerCount = mDatas.get(questionIndex).getFaqAnswerCount() + 1;
            mDatas.get(questionIndex).setFaqAnswerCount(faqAnswerCount);
            adapter.notifyDataSetChanged();
        }
    }

    @Override
    public void obBusEvent(MessageEvent event) {
        if (event.getAction().equals(Action.ALTER_FAQ_QUESTION) && event.obj != null && event.obj instanceof FAQsEntity) {
            FAQsEntity entity = (FAQsEntity) event.obj;
            if (mDatas.indexOf(entity) != -1) {
                int index = mDatas.indexOf(entity);
                mDatas.set(index, entity);
                adapter.notifyDataSetChanged();
            }
        } else if (event.getAction().equals(Action.CREATE_FAQ_QUESTION) && event.obj != null && event.obj instanceof FAQsEntity) {
            FAQsEntity entity = (FAQsEntity) event.obj;
            if (!xRecyclerView.isLoadingMoreEnabled()) {
                mDatas.add(entity);
                adapter.notifyDataSetChanged();
            }
            if (xRecyclerView.getVisibility() != View.VISIBLE)
                xRecyclerView.setVisibility(View.VISIBLE);
            if (emptyView.getVisibility() == View.VISIBLE)
                emptyView.setVisibility(View.GONE);
        } else if (event.getAction().equals(Action.COLLECTION) && event.obj != null && event.obj instanceof FAQsEntity) {
            FAQsEntity entity = (FAQsEntity) event.obj;
            if (mDatas.indexOf(entity) != -1) {
                int index = mDatas.indexOf(entity);
                mDatas.set(index, entity);
                adapter.notifyDataSetChanged();
            }
        } else if (event.getAction().equals(Action.DELETE_FAQ_QUESTION) && event.obj != null && event.obj instanceof FAQsEntity) {
            FAQsEntity entity = (FAQsEntity) event.obj;
            mDatas.remove(entity);
            adapter.notifyDataSetChanged();
            if (mDatas.size() == 0) {
                xRecyclerView.setVisibility(View.GONE);
                emptyView.setVisibility(View.VISIBLE);
            }
        } else if (event.getAction().equals(Action.CREATE_FAQ_ANSWER)) {
            int faqAnswerCount = mDatas.get(questionIndex).getFaqAnswerCount() + 1;
            mDatas.get(questionIndex).setFaqAnswerCount(faqAnswerCount);
            adapter.notifyDataSetChanged();
        }
    }

    public interface OnResponseListener {
        void getTotalCount(int totalCount);
    }
}

File: app/src/main/java/com/haoyu/app/fragment/PageAllQuestionFragment.kt
package com.haoyu.app.fragment

import android.app.AlertDialog
import android.content.Intent
import android.support.v7.widget.LinearLayoutManager
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import com.haoyu.app.activity.AppQuestionDetailActivity
import com.haoyu.app.activity.AppQuestionEditActivity
import com.haoyu.app.adapter.PageQuestionAdapter
import com.haoyu.app.base.BaseFragment
import com.haoyu.app.base.BaseResponseResult
import com.haoyu.app.basehelper.BaseRecyclerAdapter
import com.haoyu.app.dialog.MaterialDialog
import com.haoyu.app.entity.FAQsEntity
import com.haoyu.app.entity.FAQsListResult
import com.haoyu.app.entity.FollowMobileResult
import com.haoyu.app.entity.Paginator
import com.haoyu.app.lego.student.R
import com.haoyu.app.rxBus.MessageEvent
import com.haoyu.app.rxBus.RxBus
import com.haoyu.app.utils.Action
import com.haoyu.app.utils.Constants
import com.haoyu.app.utils.OkHttpClientManager
import com.haoyu.app.view.LoadFailView
import com.haoyu.app.view.LoadingView
import com.haoyu.app.view.RippleView
import com.haoyu.app.xrecyclerview.XRecyclerView
import okhttp3.Request
import java.lang.Exception
import java.util.*

/**
 * 创建日期：2018/1/15.
 * 描述:课程学习、工作坊  所有问答列表
 * 作者:xiaoma
 */
class PageAllQuestionFragment : BaseFragment(), XRecyclerView.LoadingListener {
    private lateinit var loadingView: LoadingView
    private lateinit var loadFailView: LoadFailView
    private lateinit var xRecyclerView: XRecyclerView
    private lateinit var tvEmpty: TextView
    private var type: String? = null
    private var relationId: String? = null
    private var relationType: String? = null
    private var page = 1
    private var isRefresh = false
    private var isLoadMore = false
    private val mDatas = ArrayList<FAQsEntity>()
    private lateinit var adapter: PageQuestionAdapter
    private var totalCount: Int = 0
    private var onResponseListener: OnResponseListener? = null

    fun setOnResponseListener(onResponseListener: OnResponseListener) {
        this.onResponseListener = onResponseListener
    }

    override fun createView(): Int {
        return R.layout.fragment_page_question_child
    }

    override fun initView(view: View) {
        type = arguments?.getString("type")
        relationId = arguments?.getString("relationId")
        relationType = arguments?.getString("relationType")
        loadingView = view.findViewById(R.id.loadingView)
        loadFailView = view.findViewById(R.id.loadFailView)
        xRecyclerView = view.findViewById(R.id.xRecyclerView)
        tvEmpty = view.findViewById(R.id.tv_empty)
        val layoutManager = LinearLayoutManager(context)
        layoutManager.orientation = LinearLayoutManager.VERTICAL
        xRecyclerView.layoutManager = layoutManager
        adapter = if (type != null && type == "course") {
            PageQuestionAdapter(context, mDatas, 1)
        } else {
            PageQuestionAdapter(context, mDatas, 2)
        }
        xRecyclerView.adapter = adapter
        xRecyclerView.setLoadingListener(this)
        tvEmpty.text = resources.getString(R.string.empty_questions)
        registRxBus()
    }

    override fun initData() {
        val url = Constants.OUTRT_NET + "/m/faq_question" + "?relation.id=" + relationId + "&relation.type=" + relationType + "&page=" + page + "&orders=CREATE_TIME.DESC" + "&isLoadNewstAnswer=true"
        addSubscription(OkHttpClientManager.getAsyn(context, url, object : OkHttpClientManager.ResultCallback<FAQsListResult>() {
            override fun onBefore(request: Request?) {
                if (isRefresh || isLoadMore) {
                    loadingView.visibility = View.GONE
                } else {
                    loadingView.visibility = View.VISIBLE
                }
            }

            override fun onError(request: Request?, e: Exception?) {
                loadingView.visibility = View.GONE
                when {
                    isRefresh -> xRecyclerView.refreshComplete(false)
                    isLoadMore -> {
                        page -= 1
                        xRecyclerView.loadMoreComplete(false)
                    }
                    else -> {
                        xRecyclerView.visibility = View.GONE
                        loadFailView.visibility = View.VISIBLE
                    }
                }
            }

            override fun onResponse(response: FAQsListResult?) {
                loadingView.visibility = View.GONE
                if (response?.getResponseData() != null && response.getResponseData().questions.size > 0) {
                    updateUI(response.getResponseData().questions, response.getResponseData().paginator)
                } else {
                    when {
                        isRefresh -> xRecyclerView.refreshComplete(true)
                        isLoadMore -> xRecyclerView.loadMoreComplete(true)
                        else -> {
                            xRecyclerView.visibility = View.GONE
                            tvEmpty.visibility = View.VISIBLE
                        }
                    }
                    xRecyclerView.isLoadingMoreEnabled = false
                }
            }
        }))
    }

    private fun updateUI(list: List<FAQsEntity>, paginator: Paginator?) {
        if (xRecyclerView.visibility != View.VISIBLE)
            xRecyclerView.visibility = View.VISIBLE
        if (isRefresh) {
            mDatas.clear()
            xRecyclerView.refreshComplete(true)
        } else if (isLoadMore) {
            xRecyclerView.loadMoreComplete(true)
        }
        mDatas.addAll(list)
        adapter.notifyDataSetChanged()
        if (paginator == null) {
            xRecyclerView.isLoadingMoreEnabled = false
        } else {
            totalCount = paginator.totalCount
            xRecyclerView.isLoadingMoreEnabled = paginator.hasNextPage
        }
        onResponseListener?.getTotalCount(totalCount)
    }

    override fun setListener() {
        loadFailView.setOnRetryListener { initData() }
        adapter.setCollectCallBack(object : PageQuestionAdapter.CollectCallBack {
            override fun collect(position: Int, entity: FAQsEntity) {
                collection(position)
            }

            override fun cancelCollect(position: Int, entity: FAQsEntity) {
                cancelCollection(position)
            }
        })
        adapter.setAnswerCallBack { _, entity ->
            val intent = Intent()
            intent.setClass(context, AppQuestionEditActivity::class.java)
            intent.putExtra("isAnswer", true)
            intent.putExtra("entity", entity)
            startActivity(intent)
        }
        adapter.onItemClickListener = BaseRecyclerAdapter.OnItemClickListener { _, _, _, position ->
            if (position - 1 >= 0 && position - 1 < mDatas.size) {
                val intent = Intent()
                intent.setClass(context, AppQuestionDetailActivity::class.java)
                intent.putExtra("type", type)
                intent.putExtra("entity", mDatas[position - 1])
                startActivity(intent)
            }
        }
        adapter.setOnItemLongClickListener { entity, _ ->
            if (entity.creator != null && entity.creator.id != null && entity.creator.id == userId)
                bottomDialog(entity)
        }
    }

    /**
     * 取消收藏
     *
     * @param position
     */
    private fun cancelCollection(position: Int) {
        val follow = mDatas[position].follow
        if (follow != null) {
            val url = Constants.OUTRT_NET + "/m/follow/" + follow.id
            val map = HashMap<String, String>()
            map.put("_method", "delete")
            addSubscription(OkHttpClientManager.postAsyn(context, url, object : OkHttpClientManager.ResultCallback<BaseResponseResult<*>>() {
                override fun onError(request: Request, exception: Exception) {
                    onNetWorkError()
                }

                override fun onResponse(response: BaseResponseResult<*>?) {
                    if (response?.getResponseCode() != null && response.getResponseCode() == "00") {
                        mDatas[position].follow = null
                        adapter.notifyDataSetChanged()
                        val event = MessageEvent()
                        event.action = Action.COLLECTION
                        event.obj = mDatas[position]
                        RxBus.getDefault().post(event)
                    }
                }
            }, map))
        }
    }

    /**
     * 创建收藏
     *
     * @param position
     */
    private fun collection(position: Int) {
        val url = Constants.OUTRT_NET + "/m/follow"
        val map = HashMap<String, String>()
        map.put("followEntity.id", mDatas[position].id)
        map.put("followEntity.type", "course_study_question")
        addSubscription(OkHttpClientManager.postAsyn(context, url, object : OkHttpClientManager.ResultCallback<FollowMobileResult>() {
            override fun onError(request: Request, exception: Exception) {
                onNetWorkError()
            }

            override fun onResponse(response: FollowMobileResult?) {
                if (response != null && response.responseCode != null && response.responseCode == "00") {
                    val entity = response.responseData
                    mDatas[position].follow = entity
                    adapter.notifyDataSetChanged()
                    val event = MessageEvent()
                    event.action = Action.COLLECTION
                    event.obj = mDatas[position]
                    RxBus.getDefault().post(event)
                } else {
                    response?.responseMsg?.let {
                        toast(it)
                    }
                }
            }
        }, map))
    }

    private fun bottomDialog(entity: FAQsEntity) {
        val view = LayoutInflater.from(context).inflate(R.layout.dialog_delete, LinearLayout(context), false)
        val dialog = AlertDialog.Builder(context).create()
        val rvDelete = view.findViewById<RippleView>(R.id.rv_delete)
        val rvCancel = view.findViewById<RippleView>(R.id.rv_cancel)
        val listener = RippleView.OnRippleCompleteListener {
            when (it.id) {
                R.id.rv_delete -> deleteQuestion(entity)
                R.id.rv_cancel -> {
                }
            }
            dialog.dismiss()
        }
        rvDelete.setOnRippleCompleteListener(listener)
        rvCancel.setOnRippleCompleteListener(listener)
        dialog.setCanceledOnTouchOutside(true)
        dialog.setCancelable(true)
        dialog.show()
        dialog.window?.let {
            it.setLayout(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)
            it.setWindowAnimations(R.style.dialog_anim)
            it.setContentView(view)
            it.setGravity(Gravity.BOTTOM)
        }
    }

    private fun deleteQuestion(entity: FAQsEntity) {
        val dialog = MaterialDialog(context)
        dialog.setTitle("温馨提示")
        dialog.setMessage("您确定删除此问答吗？")
        dialog.setPositiveButton("确定") { _, _ ->
            val url = Constants.OUTRT_NET + "/m/faq_question/" + entity.id
            val map = HashMap<String, String>()
            map.put("_method", "delete")
            addSubscription(OkHttpClientManager.postAsyn(context, url, object : OkHttpClientManager.ResultCallback<BaseResponseResult<*>>() {
                override fun onBefore(request: Request) {
                    showTipDialog()
                }

                override fun onError(request: Request, e: Exception) {
                    hideTipDialog()
                    onNetWorkError()
                }

                override fun onResponse(response: BaseResponseResult<*>?) {
                    hideTipDialog()
                    if (response?.getResponseCode() != null && response.getResponseCode() == "00") {
                        val event = MessageEvent()
                        event.action = Action.DELETE_FAQ_QUESTION
                        event.obj = entity
                        RxBus.getDefault().post(event)
                    }
                }
            }, map))
        }
        dialog.setCanceledOnTouchOutside(true)
        dialog.setCancelable(true)
        dialog.setNegativeButton("取消", null)
        dialog.show()
    }

    override fun onRefresh() {
        isRefresh = true
        isLoadMore = false
        page = 1
        initData()
    }

    override fun onLoadMore() {
        isRefresh = false
        isLoadMore = true
        page += 1
        initData()
    }

    override fun onEvent(event: MessageEvent) {
        if (event.getAction() == Action.CREATE_FAQ_QUESTION && event.obj != null && event.obj is FAQsEntity) {
            val entity = event.obj as FAQsEntity
            if (!xRecyclerView.isLoadingMoreEnabled) {
                mDatas.add(0, entity)
                adapter.notifyDataSetChanged()
            }
            if (xRecyclerView.visibility != View.VISIBLE) {
                xRecyclerView.visibility = View.VISIBLE
            }
            if (tvEmpty.visibility != View.GONE) {
                tvEmpty.visibility = View.GONE
            }
            onResponseListener?.getTotalCount(totalCount + 1)

        } else if (event.getAction() == Action.COLLECTION && event.obj != null && event.obj is FAQsEntity) {
            val entity = event.obj as FAQsEntity
            if (mDatas.indexOf(entity) >= 0) {
                val index = mDatas.indexOf(entity)
                mDatas[index] = entity
                adapter.notifyDataSetChanged()
            }
        } else if (event.getAction() == Action.DELETE_FAQ_QUESTION && event.obj != null && event.obj is FAQsEntity) {
            val entity = event.obj as FAQsEntity
            mDatas.remove(entity)
            adapter.notifyDataSetChanged()
            if (mDatas.size == 0) {
                xRecyclerView.visibility = View.GONE
                tvEmpty.visibility = View.VISIBLE
            }
            onResponseListener?.getTotalCount(totalCount - 1)

        } else if (event.getAction() == Action.CREATE_FAQ_ANSWER && event.obj != null && event.obj is FAQsEntity) {
            val entity = event.obj as FAQsEntity
            val index = mDatas.indexOf(entity)
            if (index >= 0) {
                val faqAnswerCount = mDatas[index].faqAnswerCount + 1
                mDatas[index].faqAnswerCount = faqAnswerCount
                adapter.notifyDataSetChanged()
            }
        }
    }

    interface OnResponseListener {
        fun getTotalCount(totalCount: Int)
    }
}